### PR TITLE
fix: prevent FP8 _scaled_mm SuperDSC aborts for Granite 8B fused-N shapes

### LINKED ISSUE #4179: SuperDSC aborts FP8 _scaled_mm at Granite fused N (6144 / 25600) and wide M=16
## What

`aten._scaled_mm` with the canonical FP8 graph (`quantize_fp8_with_scale` + `quantize_weight_fp8_with_scale` + `_scaled_mm`) fails to compile on SuperDSC for Granite 8B fused linear shapes. `dynamic=False`. Scales are graph inputs (amax is not fused into `qfp8ch`).

Works: `N ∈ {128, 1024, 4096}` and wide GEMMs at `M ∈ {1, 2, 4}`. Narrow 128-wide at M=128 is fine.

Aborts:

| Case | M | K | N | Notes |
|---|---|---|---|---|
| fused QKV | 1 | 4096 | **6144** | decode; 4096+1024+1024 |
| fused gate_up | 1 | 4096 | **25600** | 2×12800 |
| wide prefill / warmup | **16** | 4096 | 4096 | spyre-inference warmup M |
| bench prefill | 64 | 4096 | 4096 | abort without M-tile of 4 |

`insert_bmm_padding` only pads K. It does not legalize fused N or awkward M.

`qfp8wt` is a 2×64 stick layout. Work division pins the second stick dim and reduction K to `split=1` (`qfp8wt_split_domains` / `qfp8wt_matmul_k_split_domains`). Coarse tiling on main does not make N=6144 or M=16 compile as one `_scaled_mm`.

Eager `float8.to("spyre")` is the wrong layout; the graph must take FP16 `b` and assign `qfp8wt` via `quantize_weight_fp8_with_scale`. The abort is SuperDSC of that fused graph at Granite N/M.

## Why

`ibm-granite/granite-3.3-8b-instruct-FP8` (and 4.x 8B dense) uses those fused widths (hidden=4096, GQA 32/8 × 128, intermediate=12800). spyre-inference currently N-splits to `{4096, 1024, 128}` and M-tiles to 4, then graph-breaks so each tile is its own 3-op compile. That is correct but slow. Once these shapes compile as one `_scaled_mm`, we can drop the tiles.

## Repro

```python
import torch
import torch_spyre  

@torch.compile(backend="inductor", dynamic=False)
def mm(x, sa, w, sb):
    xa = torch.ops.spyre.quantize_fp8_with_scale(x, sa)
    wb = torch.ops.spyre.quantize_weight_fp8_with_scale(w, sb)
    return torch.ops.aten._scaled_mm(
        xa, wb, scale_a=sa, scale_b=sb, out_dtype=torch.float16
    )

def run(m, k, n):
    x = torch.randn(m, k, dtype=torch.float16, device="spyre")
    w = torch.randn(k, n, dtype=torch.float16, device="spyre")
    sa = (x.abs().amax() / 448).to(torch.float16).reshape(1)
    sb = (w.abs().amax() / 448).to(torch.float16).reshape(1)
    print(f"M={m} K={k} N={n}", flush=True)
    y = mm(x, sa, w, sb)
    print("ok", tuple(y.shape))

run(1, 4096, 4096)   # OK
run(1, 4096, 6144)   # abort
# run(16, 4096, 4096)  # abort
# run(1, 4096, 25600)  # abort
```

## Expected

- M=1, K=4096, N=6144 and N=25600 compile and run.
- Wide N=4096 at M=16 (and ideally prefill M=64) compiles without a host M-tile of 4.

### PR DESCRIPTION
**Pull request by Akash-Nayak**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes SuperDSC compilation failures for FP8 `_scaled_mm` operations with Granite 8B shapes (fused QKV N=6144, fused gate_up N=25600, wide prefill M=4/8/16/64).

**Fix 1 (`compute_ops.py`)** — `gen_coord_info_value` was emitting wrong constants for the QFP8WT KERNEL weight's out/N-dim (`elemArr=3` branch).
Specifically:
- `alpha[3]` was `size // 8` instead of the hardware-fixed `64`, giving an
  incorrect stride for any `n_split != 8` (e.g. `n_split=6` for N=6144 gave
  `alpha=256` instead of `64`).
- `factor` for `elem_arr_{2,1,0}` was `[64, 2, 1]` (product=64) instead of
  `[size//64, 8, 8]` (product=size), so the total element count per core was
  always 64 regardless of split.

These together caused SuperDSC to abort with `[distributeElemArrToTemporalLoops] Not enough elements to distribute`.

An alignment assert (`size % 64 == 0`) is also added, guaranteed by the invariant that N is always a multiple of 128 and `nsplits` divides `n_sticks`.

**Fix 2 (`work_division.py`)** — `_cost_model_divide_op` was only routing
`BATCH_MATMUL_OP` through `_cost_model_matmul_planner`; `BATCH_MATMUL_FP8_OP` fell through to the default `work_distribution` which picked arbitrary `n_split` values. Now both op types go through the cost model.

Regression tests added in `TestFP8IssuesVerification` covering all eight previously failing Granite 8B shapes plus baselines.

#### Which issue(s) this PR is related to:

Fixes #4179

#### Special notes for your reviewer:

The two fixes are independent and can be understood separately:
- Fix 1 touches only `gen_coord_info_value` in `compute_ops.py` (the `elemArr=3` branch for QFP8WT KERNEL out-dim tensors).
- Fix 2 is a one-line change in `_cost_model_divide_op` in `work_division.py` (add `BATCH_MATMUL_FP8_OP` to the routing condition).

#### Does this PR introduce a user-facing change?

Yes — Granite 8B FP8 inference shapes that previously caused a compile-time abort now compile and produce numerically correct results.

#### Additional note:

All 136 tests across `test_fp8_operations.py`, `test_inductor_ops.py` (fp8_scaled_mm), and `test_work_division.py` passed in local testing.

### CODE REVIEW

**Comment by Akash-Nayak** [comment]
/spyre-test force-test=hf-adapters-dev,torch-spyre-dev,spyre-inference-dev

File: tests/inductor/test_fp8_operations.py [hunk 1]
```diff
@@ -292,3 +292,95 @@ def verify_fp16_dtype(tensor):
...
+            pytest.param(1, 4096, 25600, id="fused_gate_up_n25600"),
+            pytest.param(4, 4096, 4096, id="wide_m4_n4096"),
+            pytest.param(8, 4096, 4096, id="wide_m8_n4096"),
+            pytest.param(16, 4096, 4096, id="prefill_warmup_m16_n4096"),
+            pytest.param(64, 4096, 4096, id="bench_prefill_m64_n4096"),
+            pytest.param(8, 4096, 1024, id="m8_n1024"),
+            pytest.param(8, 4096, 12800, id="m8_n12800"),
+        ],
+    )
+    def test_scaled_mm_granite_fp8_shapes(self, m, k, n):
```
Comment (joyalbin) [inline tests/inductor/test_fp8_operations.py]: Can this be added as another testcase in https://github.com/torch-spyre/torch-spyre/pull/4235/changes#diff-f8d6c2e16b0167b0bbfb49a767a60aee86f2fe84fc371796fc0e67c4d6a54bceL37 ?

Fix looks good to me, CI has to pass!!

Comment (Akash-Nayak) [reply tests/inductor/test_fp8_operations.py]: I have added it as another testcase in test_inductor_ops.py. The CI checks have passed including `/spyre-test` - https://github.com/torch-spyre/torch-spyre/pull/4235#issuecomment-5537352804.

File: torch_spyre/_inductor/work_division.py [hunk 2]
```diff
@@ -1807,12 +1807,20 @@ def factors(dim: Symbol, size: int) -> list[int]:
     if math.prod(new_splits.values()) < math.prod(splits.values()):
         if not has_qfp8wt_tensor(input_tds + [output_td]):
             return splits
-        # For QFP8WT, force k_dim = 1 regardless of core count
+        # For QFP8WT, force k_dim = 1 regardless of core count.
+        # n_dim is left as the cost model's chosen n_s; any legal divisor of
+        # n_sticks is safe because N is always a multiple of 128 (one FP8
+        # stick), so size = N // n_split = 128 * (n_sticks // n_split) is
+        # always a multiple of 64 — matching the hardware invariant
+        # P_.out_ = 64 * myCoreledOutSp from dmlite.cpp.  The assert
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/work_division.py]: Can we remove the backend references here?

Comment (Akash-Nayak) [reply torch_spyre/_inductor/work_division.py]: Done.

**Review by avery-blanchard — DISMISSED** [review]
looks good to me

File: torch_spyre/_inductor/codegen/compute_ops.py [hunk 3]
```diff
@@ -341,6 +343,40 @@ def gen_coord_info_value(
                 ],
             },
         }
+    elif is_stick_dim and is_fp8_stick and not (stick_idx == 0):
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/codegen/compute_ops.py]: https://github.com/torch-spyre/torch-spyre/pull/4483 removed this condition from main earlier in the week as it was producing invalid coordinates. Could you confirm that your tests need this?

Comment (Akash-Nayak) [reply torch_spyre/_inductor/codegen/compute_ops.py]: Thanks for flagging this @avery-blanchard . You're right, it's not needed after this condition was removed from the main. I have removed it in follow-up commit `29ea4f21`.

### What happened

When this PR was originally authored, `main` did not yet have #4483. I introduced a revised version of the `elif is_stick_dim and is_fp8_stick` branch with corrected alpha/factor values (replacing the hardcoded `size // 8` that #4483 identified as wrong).

After rebasing onto latest `main` (which includes #4483), I have verified by adding debug prints to the `batchmatmulfp8` and `else/elemArr=2` branches and running `test_fp8_scaled_mm_granite_fp8_shapes_decode_m1_n4096_baseline`. The `qfp8wt` outer-stick dim falls through cleanly to `elemArr=2` without the dedicated `elif` branch:

```
[DEBUG] else/elemArr=2 branch fired: size=128, nsplits=32, is_stick_dim=True, tensor_idx=0, opfunc=qfp8wt
[DEBUG] else/elemArr=2 branch fired: size=4096, nsplits=1, is_stick_dim=True, tensor_idx=1, opfunc=qfp8wt
[DEBUG] else/elemArr=2 branch fired: size=128, nsplits=32, is_stick_dim=True, tensor_idx=1, opfunc=qfp8wt
[DEBUG] batchmatmulfp8 branch fired: size=4096, nsplits=1, tensor_idx=0, opfunc=batchmatmulfp8
[DEBUG] else/elemArr=2 branch fired: size=4096, nsplits=1, is_stick_dim=True, tensor_idx=1, opfunc=batchmatmulfp8
```

### What was removed in `29ea4f21`

- `is_fp8_stick` / `stick_idx` params from `gen_coord_info_value`
- The entire `elif is_stick_dim and is_fp8_stick and not (stick_idx == 0)` block (`elemArr=3`)
- The `is_fp8` / `st_idx` computation and call-site args in `_build_coord_info`

### Test results (on this branch, after removal)

| Suite | Result |
|---|---|
| `test_inductor_ops.py -k fp8` (58 tests) | ✅ all passed |
| `test_fp8_operations.py` (68 tests) | ✅ all passed |

**Review by avery-blanchard — APPROVED** [review]
lgtm, thanks!
